### PR TITLE
Improve resource monitor errors

### LINKED ISSUE #4869: [FEATURE] Improve resource manager error messaging
**Is your feature request related to a problem?**
When running into memory contention in a node, we run a resource manager check before query execution, and on failure return an error: `Insufficient resources to run the query, quit.` This error lacks context on what resource is insufficient & it isn't clear that this happens before query execution instead of being due to the query.

This ultimately directs to the ResourceMonitor check being an isHealthy boolean:
https://github.com/opensearch-project/sql/blob/b88bf567ad20dadc5742d2cedc9e8c7f0427f617/core/src/main/java/org/opensearch/sql/monitor/ResourceMonitor.java#L13-L18

Which doesn't give any context to the caller on *why* it's unhealthy.

**What solution would you like?**
Instead of a boolean, a resource monitor could return some sort of status object which indicates the type of threshold being broken.

```java
public abstract ResourceStatus checkResource();
```

This status might support methods like `isHealthy` and `getDescription`, would give context to the caller that can be presented to the user. With some error message wrangling, the current `insufficient resources` error can be reworked into a 429 with an error like:

```
Unable to begin processing the query due to insufficient memory. Check the cluster memory pressure.
To raise the query memory cutoff, update the `plugins.query.memory_limit` setting (current: 85%).
```

**What alternatives have you considered?**
N/A

**Do you have any additional context?**
```py
# source=big5 | head;
TransportError(500, 'RuntimeException', {
  'error': {
    'reason': 'There was internal problem at backend',
    'details': 'java.sql.SQLException: exception while executing query: insufficient resources to run the query, quit.',
    'type': 'RuntimeException'
  }, 'status': 500
})
```

Internal ticket (for AWS employees): https://t.corp.amazon.com/V2003709229

### PR DESCRIPTION
**Pull request by Swiddis**
### Description
Quick update for resource management errors.

Technically related to work on #4919 but these changes don't involve the framework, this just adds some resource monitoring infrastructure that didn't exist before.

Before:
```json
{
  "error": {
    "reason": "There was internal problem at backend",
    "details": "java.sql.SQLException: exception while executing query: insufficient resources to run the query, quit.",
    "type": "RuntimeException"
  },
  "status": 500
}
```

After:
```json
{
  "error": {
    "reason": "There was internal problem at backend",
    "details": "Insufficient resources to start query: Memory usage exceeds limit: 74510336 bytes used, 21474836 bytes limit (current: 71.1MB, limit: 20.5MB, usage: 347.0%). To increase the limit, adjust the 'plugins.query.memory_limit' setting (default: 85%).",
    "type": "RuntimeException"
  },
  "status": 500
}
```

### Related Issues
Resolves #4869

### Check List
- [X] New functionality includes testing.
- [ ] New functionality has been documented.
 - [ ] New functionality has javadoc added.
 - [ ] New functionality has a user manual doc added.
- [ ] New PPL command [checklist](https://github.com/opensearch-project/sql/blob/main/docs/dev/ppl-commands.md) all confirmed.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff` or `-s`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: opensearch/src/main/java/org/opensearch/sql/opensearch/monitor/OpenSearchResourceMonitor.java [hunk 1]
```diff
@@ -60,4 +60,29 @@ public boolean isHealthy() {
...
+  @Override
+  public org.opensearch.sql.monitor.ResourceStatus getStatus() {
+    try {
+      ByteSizeValue limit = settings.getSettingValue(Settings.Key.QUERY_MEMORY_LIMIT);
+      if (limit == null) {
+        // undefined, be always healthy
+        return org.opensearch.sql.monitor.ResourceStatus.healthy(
+            org.opensearch.sql.monitor.ResourceStatus.ResourceType.MEMORY);
+      }
+      return memoryMonitor.getMemoryStatus(limit.getBytes());
```
Comment (penghuo) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/monitor/OpenSearchResourceMonitor.java]: bypasses the retry mechanism that isHealthy() uses. is it intentional? The
existing isHealthy() wraps memoryMonitor.isMemoryHealthy() with Retry.decorateSupplier(retry, ...) (3 attempts, exponential backoff) to tolerate transient MemoryUsageExceedException.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/executor/protector/ResourceMonitorPlan.java [hunk 2]
```diff
@@ -43,8 +43,14 @@ public <R, C> R accept(PhysicalPlanNodeVisitor<R, C> visitor, C context) {
...
   public void open() {
-    if (!this.monitor.isHealthy()) {
-      throw new IllegalStateException("insufficient resources to run the query, quit.");
+    org.opensearch.sql.monitor.ResourceStatus status = this.monitor.getStatus();
+    if (!status.isHealthy()) {
+      throw new IllegalStateException(
+          String.format(
+              "Insufficient resources to start query: %s. "
+                  + "To increase the limit, adjust the 'plugins.query.memory_limit' setting "
+                  + "(default: 85%%).",
```
Comment (penghuo) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/executor/protector/ResourceMonitorPlan.java]: nit: The error message hardcodes plugins.query.memory_limit and
(default: 85%). If the setting name or default changes, these messages go stale. Was a constant or Settings.Key.QUERY_MEMORY_LIMIT.getKeyValue() considered for the setting name?

File: opensearch/src/test/java/org/opensearch/sql/opensearch/monitor/OpenSearchResourceMonitorTest.java [hunk 3]
```diff
@@ -76,4 +78,68 @@ void healthyWithRetry() {
...
+        .thenReturn(healthyStatus);
+
+    OpenSearchResourceMonitor resourceMonitor =
+        new OpenSearchResourceMonitor(settings, memoryMonitor);
+    ResourceStatus status = resourceMonitor.getStatus();
+    assertTrue(status.isHealthy());
+    assertEquals("Memory is healthy", status.getDescription());
+    // Should call twice: first unhealthy, then healthy
+    verify(memoryMonitor, times(2)).getMemoryStatus(anyLong());
+  }
```
[the remark this answers is not in the document]
Comment (Swiddis) [reply opensearch/src/test/java/org/opensearch/sql/opensearch/monitor/OpenSearchResourceMonitorTest.java:144]: Status edge cases are separately handled in ResourceStatusTest.

**Review by ahkcs — APPROVED** [review]
